aten/src/ATen/Utils.h

```diff
@@ -31,27 +31,38 @@ static inline T* checked_cast_storage(Base* expr, const char * name, int pos) {
 }
 
 template <typename T, typename Base>
-inline T* checked_cast_tensor(Base* expr, const char * name, int pos, bool allowNull) {
+inline T* checked_cast_tensor(Base* expr, const char * name, int pos, bool allowNull, Backend backend, ScalarType scalar_type) {
   if(allowNull && expr == UndefinedTensor::singleton()) {
     return nullptr;
   }
-  if (typeid(*expr) != typeid(T))
-    AT_ERROR("Expected object of type ", T::typeString(), " but found type ", expr->type().toString(),
+  if (expr->type().backend() != backend) {
+    AT_ERROR("Expected object of backend ", backend, " but got backend ", expr->type().backend(),
+             " for argument #", pos, " '", name, "'");
+  }
+  if (expr->type().scalarType() != scalar_type) {
+    AT_ERROR("Expected object of scalar type ", scalar_type, " but got scalar type ", expr->type().scalarType(),
              " for argument #", pos, " '", name, "'");
+  }
+  // This is going away soon; delete this when we remove the subtypes of
+  // TensorImpl (so that we can eliminate the T template parameter)
+  if (typeid(*expr) != typeid(T)) {
+    AT_ERROR("Expected object of RAII type ", typeid(T).name(), " but found type ", typeid(*expr).name(),
+             " for argument #", pos, " '", name, "'");
+  }
   return static_cast<T*>(expr);
 }
 
 // Converts a TensorList (i.e. ArrayRef<Tensor> to the underlying TH* Tensor Pointer)
 template <typename T, typename TBase, typename TH>
-static inline std::vector<TH*> tensor_list_checked_cast(ArrayRef<TBase> tensors, const char * name, int pos) {
```

Comment:
you don't seem to actually need the new parameters here.

Reply:
and it seems like generated call sites are different, so there's not really a benefit to passing them in.

Reply:
I'll need it soon, as soon as I get rid of the derived types I'll need the parameters because I won't have T available anymore.